### PR TITLE
schedule-free Prodigy

### PR DESCRIPTION
**Pull request by dxqbYD**
I've been testing your other contribution of schedule-free optimizers recently on diffusion models, and it seems to work great: It converges fast and then the samples remain stable, even if you continue to train for a while, without overfitting.
https://arxiv.org/abs/2405.15682

This is a draft of a schedule-free Prodigy.

I've been testing mostly a combination of schedule-free Prodigy and these PRs, https://github.com/konstmish/prodigy/pull/22 https://github.com/konstmish/prodigy/pull/23 https://github.com/konstmish/prodigy/pull/25.
You can find the combination here:
https://github.com/dxqbYD/prodigy/tree/sliced_diced_factored_schedulefree_with_defaults

This draft PR is a reduced version of only-schedulefree Prodigy, but hasn't been tested much.

Obviously, when combining so many concepts you never know how they affect each other. I hope eventually the original authors find the time to combine their amazing contributions themselves and can provide the mathematical and experimental proof that it works!

### CODE REVIEW

**Comment by LoganBooker** [comment]
@dxqbYD This is kind of crazy -- I'd been working on incorporating schedule-free with Prodigy myself for the last week or two. Like you, I'm not sure if the maths and theory all line up, but I've also had great results with it training SDXL LoRAs. Additionally, I'd already implemented a version of the 0-tensor optimisation, but all I did was look for zero weights at initialisation, skip storing `p0` altogether, and then just do `(grad * -p.data).sum()` (or something to that effect) for the numerator updates.

Seeing this pull request, I finally decided to upload and publish my attempt: https://github.com/LoganBooker/prodigy-plus-schedule-free. This also integrates your optimisation pull requests (with some tweaks of my own) -- the slicing approach to the parameters is great, by the way. My version also allows for per-group `d` and `d_numerator` calculations (adapted from here: https://github.com/konstmish/prodigy/pull/20).

It looks like the authors will integrate at least some of your changes soon, so good work!

**Comment by dxqbYD** [comment]
> @dxqbYD This is kind of crazy -- I'd been working on incorporating schedule-free with Prodigy myself for the last week or two. Like you, I'm not sure if the maths and theory all line up, but I've also had great results with it training SDXL LoRAs. Additionally, I'd already implemented a version of the 0-tensor optimisation, but all I did was look for zero weights at initialisation, skip storing `p0` altogether, and then just do `(grad * -p.data).sum()` (or something to that effect) for the numerator updates.
> 
> Seeing this pull request, I finally decided to upload and publish my attempt: https://github.com/LoganBooker/prodigy-plus-schedule-free. This also integrates your optimisation pull requests (with some tweaks of my own) -- the slicing approach to the parameters is great, by the way. My version also allows for per-group `d` and `d_numerator` calculations (adapted from here: #20).
> 
> It looks like the authors will integrate at least some of your changes soon, so good work!

That's great!
We have a little group of testers on the OneTrainer discord, you are welcome to join. A few people have been testing my PRs there. Feedback on schedule-free and slicing has been only positive so far, but they have found some issues with the Adafactor-extensions (just commented here: https://github.com/konstmish/prodigy/pull/25#issuecomment-2454112866).
You'll find the channel in #development

Someone there also tried to merge https://github.com/konstmish/prodigy/pull/20 but wasn't successful so far. He might be interested in your work.

**Comment by dxqbYD** [comment]
>  https://github.com/LoganBooker/prodigy-plus-schedule-free. 

just had a quick look at your code. You are right that the complexity of all this does call for a re-factor into multiple functions and such, and maybe add support for for-each and a fused backpass.
I've refrained from doing so in my code so far, so that changes can still be easily tracked between this repo and the PRs

**Comment by LoganBooker** [comment]
> Someone there also tried to merge https://github.com/konstmish/prodigy/pull/20 but wasn't successful so far. He might be interested in your work.
Interesting. I did clean up a lot of redundant stuff in the PR for my version and haven't had any issues so far.

What I will say is that the default Prodigy is more robust to lopsided `d` estimates by essentially averaging them across parameter groups. The best example I can give is training SDXL LoRAs with both text encoders and the Unet. The TEs stabilise quite quickly, but the Unet grows very slowly and never really stabilises -- they end up balancing each out out in practice. And hence why combining Prodigy with cosine works so well for this purpose.

And thanks for the tip about the factored version. I suspect as you've said it's a scaling or precision issue.

**Comment by dxqbYD** [comment]
> @dxqbYD This is kind of crazy -- I'd been working on incorporating schedule-free with Prodigy myself for the last week or two. Like you, I'm not sure if the maths and theory all line up, but I've also had great results with it training SDXL LoRAs. Additionally, I'd already implemented a version of the 0-tensor optimisation, but all I did was look for zero weights at initialisation, skip storing `p0` altogether, and then just do `(grad * -p.data).sum()` (or something to that effect) for the numerator updates.
> 
> Seeing this pull request, I finally decided to upload and publish my attempt: https://github.com/LoganBooker/prodigy-plus-schedule-free. This also integrates your optimisation pull requests (with some tweaks of my own) -- the slicing approach to the parameters is great, by the way. My version also allows for per-group `d` and `d_numerator` calculations (adapted from here: #20).
> 
> It looks like the authors will integrate at least some of your changes soon, so good work!

Had a chance now to have a closer look. You have been clearly following the work here from the beginning.

I think we should give the original authors a chance to have a look at the PRs and merge what they want to have in the original repo, not run out the gates with a fork.

**Comment by LoganBooker** [comment]
Hey dxqbYD, you're absolutely right I've been following your updates! I can't stress enough how crazy it was to see your pull request for the zero-tensor optimisation. I was quite literally looking to fork and submit my own pull request for my attempt (as described in my first comment), but your approach was just flat-out better in terms of handling the problem more generally. I should stress I already had a schedule-free version running locally at this stage.

In terms of forks, etc. -- I don't see an issue with them coexisting. I don't think anyone will be using my version any time soon, it's far too experimental (for example, it uses Adam-atan2 for updates, which I'm still not 100% sure work with Prodigy and SF correctly), and I'd urge everyone to use the main Prodigy repo and cherry-pick your requests.

**Comment by TheForgotten69** [comment]
Using this version exclusively for Flux and it's working extremely great so far. Much better than prodigy or its factored version - Thanks a lot @dxqbYD !

**Comment by dxqbYD** [comment]
> Any test on Dit Model like Flux or SD3.5? My own use of prodigy doesn't perform very well on the DiT model.

On Flux LoRA, yes.
If anyone wants to explore further how schedule-free actually behaves on different models, I'd recommend getting a trainer such as OneTrainer to log `d_hat` and some gradient norm value to tensorboard, to be able to see what schedule-free actually does. Now it is kind-of a black box and you don't really see when it has decided to "stop training".

image

rms is this line: https://github.com/konstmish/prodigy/blob/a88d2879cc9cd0a112bdc1627afcd2fdf7928f9d/prodigyopt/prodigy.py#L114

**Comment by konstmish** [comment]
I quickly checked the implementation and I don't see any support for the eval mode (see [ScheduleFree repo](https://github.com/facebookresearch/schedule_free?tab=readme-ov-file#how-to-use) for the details). Did the schedule free method work for you without them? Did you run tests on networks with batch norm? For reference, [here is the eval implementation](https://github.com/facebookresearch/schedule_free/blob/49f34e0cf3ff5646ed2dd3f664d012d71d119c97/schedulefree/adamw_schedulefree.py#L78) for AdamW.

**Comment by dxqbYD** [comment]
> I quickly checked the implementation and I don't see any support for the eval mode (see [ScheduleFree repo](https://github.com/facebookresearch/schedule_free?tab=readme-ov-file#how-to-use) for the details). Did the schedule free method work for you without them? Did you run tests on networks with batch norm? For reference, [here is the eval implementation](https://github.com/facebookresearch/schedule_free/blob/49f34e0cf3ff5646ed2dd3f664d012d71d119c97/schedulefree/adamw_schedulefree.py#L78) for ScheduleFree AdamW.

I think I missed that and it worked.
This draft PR was more like a fun experiment than a proposal to merge. I'm sure for a merge you want to make sure that the math still works in combination, which I am not able to do.
